lib/nostrum/bot.ex

```diff
@@ -316,9 +316,47 @@ defmodule Nostrum.Bot do
   @doc false
   def set_bot_pid(pid), do: Process.put(:nostrum_bot_pid, pid)
 
-  defp set_bot(%{pid: pid, bot_options: _bot_options, name: name}) do
-    set_bot_pid(pid)
-    set_bot_name(name)
+  defp set_bot(%{pid: pid, bot_options: _bot_options, name: name}), do: set_bot(pid, name)
+  defp set_bot(pid, name), do: {set_bot_pid(pid), set_bot_name(name)}
+
+  @doc """
+  Call a function with the context of the provided bot
+
+  ```elixir
+  Nostrum.Bot.with_bot(ReticentBot, fn ->
+    Nostrum.Api.Message.create(channel_id, "Hi...")
+  end)
+
+  Nostrum.Bot.with_bot(SillyGoofyBot, fn ->
+    Nostrum.Api.Message.create(channel_id, "Heyyy :)")
+  end, :keep)
+
+  # Context is still set to SillyGoofyBot
+  Nostrum.Api.Message.create(channel_id, "Me again :)")
+  ```
+
+  ## Parameters
+
+  - `name` - Name of the bot to run the function as
+  - `function` - Zero arity function to execute
+  - `next` - What to do with the process's bot context after the function
+    - `:reset` - Reset the values to what they were before (default)
+    - `:clear` - Set the values to `nil`
+    - `:keep` - Keep the provided bot's context in this process's dictionary
+  """
+  @spec with_bot(name(), (-> any()), :reset | :clear | :keep) :: any()
+  def with_bot(name, function, next \\ :reset) do
+    {prev_pid, prev_name} = set_bot(fetch_bot_pid(name), name)
+
+    result = function.()
+
+    case next do
+      :reset -> set_bot(prev_pid, prev_name)
+      :clear -> set_bot(nil, nil)
+      :keep -> :noop
+    end
+
+    result
```

Comment:
```suggestion
try do
  function.()
after
  case next do
     :reset -> set_bot(prev_pid, prev_name)
     :clear -> set_bot(nil, nil)
     :keep -> :noop
  end
end
```

Just because the fn can raise and we probably should still do what's requested instead of what'd be `:keep` if an exception was caught further up.